yt/utilities/configuration_tree.py

```diff
@@ -0,0 +1,180 @@
+class ConfigNode:
+    def __init__(self, key, parent=None):
+        self.key = key
+        self.children = {}
+        self.parent = parent
+
+    def add(self, key, child):
+        self.children[key] = child
+        child.parent = self
+
+    def update(self, other, extraData=None):
+        def _helper(d, keys):
+            for key, val in d.items():
+                new_keys = keys + [key]
+                if isinstance(val, dict):
+                    _helper(val, new_keys)
+                else:
+                    self.upsert_from_list(new_keys, val, extraData)
+
+        _helper(other, [])
+
+    def get_child(self, key, constructor=None):
+        if key in self.children:
+            child = self.children[key]
+        elif constructor is not None:
+            child = self.children[key] = constructor()
+        else:
+            raise KeyError(f"Cannot get key {key}")
+        return child
+
+    def add_child(self, key):
+        self.get_child(key, lambda: ConfigNode(key, parent=self))
+
+    def remove_child(self, key):
+        self.children.pop(key)
+
+    def upsert_from_list(self, keys, value, extraData=None):
+        key, *next_keys = keys
+        if len(next_keys) == 0:  # reach the end of the upsert
+            leaf = self.get_child(
+                key,
+                lambda: ConfigLeaf(key, parent=self, value=value, extraData=extraData),
+            )
+            leaf.value = value
+            leaf.extraData = extraData
+            if not isinstance(leaf, ConfigLeaf):
+                raise RuntimeError(f"Expected a ConfigLeaf, got {leaf}!")
```

Comment:
The `upsert_from_list` function is called when setting an item in the config, which can be called directly by the user using `ytcfg.set("yt", "foo", "bar", "baz")`.
With the config
```
[yt.test.foo]
bar = 42
```
a `RuntimeError` would be triggered by `ytcfg.set("yt", "test", "foo", "some_value")` because foo is not an option (it is a section).

Reply:
Either your explanation has inconsistency or I'm not following :/
Shouldn't "baz" be the value instead of `42`?

Reply:
I can't understand my own initial comment, so let's try again. Essentially, these functions are exposed almost directly to the user (through the `YTConfig` object, which just queries the config tree), so I don't want to have it throw assertion errors, and rather something slightly more useful (though not _very_ useful).

Reply:
ok, makes a lot of sense. I didn't realise this code was so close to the surface.